### ISSUE TITLE
Unable to open M1 distribution

### ISSUE THREAD

**Issue by kennycason** [libgdx/gdx-liftoff#194]
Currently having an issue opening the Mac M1 release. 

Version: `gdx-liftoff-1.12.1.13-macM1`

Command to open: `open gdx-liftoff.app`

The error:
```
The application cannot be opened for an unexpected reason, error=Error Domain=RBSRequestErrorDomain Code=5 "Launch failed." UserInfo={NSLocalizedFailureReason=Launch failed., NSUnderlyingError=0x600003aa92c0 {Error Domain=NSPOSIXErrorDomain Code=111 "Unknown error: 111" UserInfo={NSLocalizedDescription=Launchd job spawn failed}}}
```

**Comment by tommyettinger** [comment]
Dang. @fourlastor , might need some help here... I have no idea what that error is and I have no way of testing on any kind of Mac, M1 or not. It looks related to Construo, since that's what built the .app ...

**Comment by fourlastor** [comment]
I wonder if this is a signing issue? I assume liftoff isn't signed, so the first thing I'd try is `xattr -cr gdx-liftoff.app` (which clears the attributes on the folder, included the quarantine attribute).

In that case, to avoid the quarantine, people could download liftoff with `curl -L https://github.com/libgdx/gdx-liftoff/releases/download/v1.12.1.14/gdx-liftoff-macM1.zip` (or whichever url the latest release points at), as curl doesn't add the quarantine to files it downloads

**Comment by fourlastor** [comment]
Ah wait, I just tried to download both the linux and the mac version, and their executables don't have the executable permission :grimacing: 

```
➜  MacOS ll
total 23M
-rw-r--r-- 1 daniele daniele  149 Aug 16 21:33 config.json
-rw-r--r-- 1 daniele daniele 626K Aug 16 21:33 gdx-liftoff
-rw-r--r-- 1 daniele daniele  23M Aug 16 21:33 gdx-liftoff-1.12.1.14.jar
drwxr-xr-x 5 daniele daniele 4.0K Aug 16 21:33 jdk
➜  gdx-liftoff-1.12.1.14-linuxX64 ll
total 23M
-rw-r--r-- 1 daniele daniele  149 Aug 16 21:34 config.json
-rw-r--r-- 1 daniele daniele 711K Aug 16 21:34 gdx-liftoff
-rw-r--r-- 1 daniele daniele  23M Aug 16 21:34 gdx-liftoff-1.12.1.14.jar
drwxr-xr-x 5 daniele daniele 4.0K Aug 16 21:34 jdk
```

This can be verified by running `chmod +x gdx-liftoff.app/Contents/MacOS/gdx-liftoff`

Did you package these on windows? That might be an issue with gradle running on Windows and not managing to change the executable permission

**Comment by tommyettinger** [comment]
Indeed, I did package them on Windows, since cross-compilation is supposed to be something Construo can do... heh.

**Comment by JojoIce** [comment]
> This can be verified by running `chmod +x gdx-liftoff.app/Contents/MacOS/gdx-liftoff`

I still get a popup: "“gdx-liftoff.app” is damaged and can’t be opened. You should move it to the Bin." when attempting to open the app.

However starting from command-line with java -jar gdx-liftoff-1.12.1.14.jar from within the app folder works for me. Not familiar with Construo so can't really debug, but still looks like something is off with the packaging.

> I think .zip format may not always preserve chmod flags...

The app seems to be ok as in it is executable, but the files within the app (app for mac is a slightly special folder) seems to have the wrong rights, but for me not the (only?) problem.

**Comment by JojoIce** [comment]
Well, this peaked my interest, and I saw that the project that i managed to create using the command-line (1.12.1.14) had the task to create a construo packaged Mac distro. Happy to report that it works just fine on the Mac M1 that I tested it on! A proper mac app right away, very nice!

One small issue is that the folder inside of the zip that is created by construo was named projectname-$projectVersion-macM1, where "projectname" is just whatever project name you give it and the rest literally what I wrote, $ and all. The version for that app is 1.0 no matter what you set the gradle.properties projectVersion to. The zip-file does not have a version number in the name, just the project name and platform. The app had the projectname.app just as expected. The gradle task was executed on the same Mac M1 as the app was tested on.

Under lwjgl3/build/distributions/ built old-fashioned way the file will be called "lwjgl3-$projectVersion.zip"

Under lwjgl3/build/libs/ it will look as expected with proj name and version .jar

**Comment by JojoIce** [comment]
Oh, I have been doing other things and looking into this in short bursts. I tried what fourlastor mentioned. First one and then the other, but not on the same download. Doing both allows me to execute the app, so:
xattr -cr gdx-liftoff.app
chmod +x gdx-liftoff.app/Contents/MacOS/gdx-liftoff
Fixed it for me.

**Comment by tommyettinger** [comment]
I think I can look into the dollar-sign problem from here on Windows; it's challenging to get that right because there are multiple levels of escaping for all Kotlin string interpolation in Liftoff. There's the template itself, which is typically a triple-quoted Kotlin string. Interpolations in that look things up in Liftoff. Or, you can insert a dollar sign in the template, which is cumbersome, and have that resolve to an interpolation when the user opens their project. You might also need to enter an escaped dollar sign in the template so the user gets an escaped dollar sign and not an interpolation. Blurgh.

**Comment by tommyettinger** [comment]
It turned out to be a much simpler quoting issue -- I (or fourlastor, maybe?) had used single quotes when the version was hard-coded initially, and then I changed that to use `projectVersion`, but forgot to make it use double quotes. Single-quoted strings in Groovy don't perform string interpolation, but double-quoted ones do.

**Comment by badlogic** [comment]
@tommyettinger the issue here is that ZIP files do not retain file permissions. You need to distribute binaries for Linux and macOS using e.g. tar.gz which will work on Linux and macOS.

**Comment by fourlastor** [comment]
Yup! I'm planning on copying packr approach https://github.com/fourlastor-alexandria/construo/issues/64

.zip in my experience retains permissions but not when you unzip it on windows

**Comment by badlogic** [comment]
Ah, yes, you are right. Silly old Mario brain.

**Comment by PromanSEW** [comment]
Also can't open app on MacBook Pro M2, Mac OS Sonoma 14.6.1
Just "Unable to open program 'gdx-liftoff'"

**Comment by tommyettinger** [comment]
Yep, we know now that the MacOS native package won't run out of the box, but the .jar will. Thank you. I'm pretty sure fourlastor is working on a new version of Construo, which Liftoff will use and projects it generates will use, that produces runnable .app files on MacOS.

**Comment by fourlastor** [comment]
I think that's definitely something worth trying, but as of now, it will require a mac to sign the app, as construo itself isn't able to sign the app yet (it's an [open issue](https://github.com/fourlastor-alexandria/construo/issues/36), but not having an Apple computer myself to test it, it's going to take a while if I need to implement it myself).

On the other hand, given the scope of liftoff (open liftoff, generate project, close liftoff), it might not be worth the time spent trying to make it work.
